### ISSUE TITLE
beforeHandler and schema validation order

### ISSUE THREAD

**Issue by peteyycz** [fastify/fastify#1105]
Currently all schema validation (like body, params, query) are performed **before** the `beforeHandler` hook is called, therefore authentication cannot be performed before the schema validation which might expose some sensitive information about the schemas of protected resources.

**Comment by delvedor** [comment]
Hi! As you can see in our [lifecycle](https://github.com/fastify/fastify/blob/master/docs/Lifecycle.md) documentation, the validation always happen before `preHandler` hooks and the `beforeHandler` (which is just a route specific `preHandler`).
If needed you can run your authentication logic in the `onRequest` hook :)

**Comment by mcollina** [comment]
I think we should swap that order, and execute `beforeHandler` _after_ validation, or add a new hook.

**Comment by peteyycz** [comment]
What I'm trying to achieve here is the following:

```javascript
  fastify.route({
    method: "POST",
    url: "/admins",
    beforeHandler: fastify.jwtAuth({ role: USER_ROLES.ADMIN }),
    schema: {
      body: {
        type: "object",
        required: ["password", "email"],
        properties: {
          password: {
            type: "string",
            minLength: 10,
          },
          email: {
            type: "string",
          },
        },
      },
    },
    handler: async (request, reply) => {
      console.log(1);
      const { email, password } = request.body;
      try {
        await fastify.models.admin.createAdmin({ email, password });
      } catch (ex) {
        if (ex.code === "23505") {
          return reply.code(409).send({
            message: "email is already registered",
          });
        }
        throw ex;
      }
      return reply.code(201).send();
    },
  });
```

**Comment by jsumners** [comment]
A new hook would be least disruptive. `beforeValidation` seems appropriate.

**Comment by cemremengu** [comment]
I think another hook may cause a hook bloating though and (maybe) leak abstraction. The framework should also prevent `[exposing] some sensitive information about the schemas of protected resources.` by default and not leave it to user understanding of the framework.

**Comment by delvedor** [comment]
@jsumners me and @mcollina where discussing the same thing, we should add a new hook that runs before the authentication bu after the middlewares.
I'll continue the work started in https://github.com/fastify/fastify/pull/955.

**Comment by cemremengu** [comment]
Landed in #1153
